### PR TITLE
Never use the entire span of a function declaration or function expression when reporting a checker error.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -218,32 +218,39 @@ module ts {
...
+            length: span.length,
             code: messageChain.code,
             category: messageChain.category,
             messageText: messageChain.next ? messageChain : messageChain.messageText
         };
     }
 
-    export function getErrorSpanForNode(node: Node): Node {
-        var errorSpan: Node;
+    interface FileTextRange extends TextRange {
```
Comment (JsonFreeman) [inline src/compiler/utilities.ts]: Remove

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -254,16 +261,35 @@ module ts {
             case SyntaxKind.ModuleDeclaration:
             case SyntaxKind.EnumDeclaration:
             case SyntaxKind.EnumMember:
-                errorSpan = (<Declaration>node).name;
+            case SyntaxKind.FunctionDeclaration:
+            case SyntaxKind.FunctionExpression:
```
Comment (JsonFreeman) [inline src/compiler/utilities.ts]: Remove this case label

File: src/compiler/utilities.ts [hunk 3]
```diff
@@ -254,16 +261,35 @@ module ts {
...
+            }
+        }
 
         // We now have the ideal error span, but it may be a node that is optional and absent
         // (e.g. the name of a function expression), in which case errorSpan will be undefined.
         // Alternatively, it might be required and missing (e.g. the name of a module), in which
         // case its pos will equal its end (length 0). In either of these cases, we should fall
         // back to the original node that the error was issued on.
-        return errorSpan && errorSpan.pos < errorSpan.end ? errorSpan : node;
+        errorNode = nodeIsMissing(errorNode) ? node : errorNode;
```
Comment (JsonFreeman) [inline src/compiler/utilities.ts]: Maybe just a span of length 1 on the current position.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -11844,9 +11836,7 @@ module ts {
         function grammarErrorOnNode(node: Node, message: DiagnosticMessage, arg0?: any, arg1?: any, arg2?: any): boolean {
             var sourceFile = getSourceFileOfNode(node);
             if (!hasParseDiagnostics(sourceFile)) {
-                var span = getErrorSpanForNode(node);
-                var start = span.end > span.pos ? skipTrivia(sourceFile.text, span.pos) : span.pos;
-                diagnostics.add(createFileDiagnostic(sourceFile, start, span.end - start, message, arg0, arg1, arg2));
+                diagnostics.add(createDiagnosticForNode(node, message, arg0, arg1, arg2));
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Maybe call this createNodeDiagnostic

**Comment by JsonFreeman** [comment]
:+1:

File: src/compiler/utilities.ts (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/compiler/utilities.ts]: D'oh!

File: src/compiler/utilities.ts (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/compiler/utilities.ts]: Done.

File: src/compiler/utilities.ts (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/compiler/utilities.ts]: I'm going to try that out.

**Pull request merged**